### PR TITLE
fix: Windows extra file

### LINKED ISSUE #421: [BUG] Pyproject.toml workaround can make a file
The following pyproject.toml on Windows:

```toml
requires = [
  "setuptools_scm[toml]>=4.1.2"
]
```

produces a file

```
4.1.2
```

which breaks the setuptools_scm versioning, since it's now a dirty directory. Illustrated in #420. I think it used to work originally, as I made a few releases after the Windows 3.5 patch initially went in.

This _only_ breaks if there is a bracketed requirement.

### PR DESCRIPTION
**Pull request by henryiii**
Closes #421

### CODE REVIEW

**Comment by joerick** [comment]
Looks to me that `pip install` line should be a `call`, not a `shell`. The subprocess module should then take care of the escaping.

**Comment by henryiii** [comment]
Okay, I'll try that! FYI, `]>=` seems to trigger the issue, while `] >=`doesn't.

**Comment by henryiii** [comment]
Call seems to do the same thing. Replacing `"]"` with `"] "` fixed the issue. I'm not familiar enough with cmd to know why this is breaking/working. One thing I do know; you do have to explicitly add the quotes on some shells (zsh, most famously) if you use square brackets. Don't know why call is affected, though.

**Comment by Czaki** [comment]
Did you check content of this strange file?

**Comment by henryiii** [comment]
Nope, it's always in CI, so I don't know what's in it. I could cat it, but not sure I know how to do that on Windows.

**Comment by Czaki** [comment]
> Nope, it's always in CI, so I don't know what's in it. I could cat it, but not sure I know how to do that on Windows.

I thinkt that you could do this with conditional step  https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions

in this PR you could add code which will check if such file

Maybe double quotes may help: https://www.robvanderwoude.com/escapechars.php

**Comment by YannickJadoul** [comment]
Would there be a chance to run it on a more sane shell, like powershell? But I guess that's maybe not included on ever CI service :-(

**Comment by henryiii** [comment]
I don't think you can avoid having powershell.

**Comment by henryiii** [comment]
At the same time, I'm not sure how you might change subprocess's calling system.

**Comment by YannickJadoul** [comment]
> At the same time, I'm not sure how you might change subprocess's calling system.

I thought you could pass a string as argument to the `powershell` command?

**Comment by joerick** [comment]
I think perhaps the most robust solution here would be to remove `shell=True` from `call()`. Perhaps we could figure out a solution to the PATH resolution problem - resolve programs on PATH ourselves? Or maybe using Windows `where` command?

**Comment by joerick** [comment]
ooh- another idea, sidestep the command line completely - write those pep518 requirements to a temp file and do `pip install -r tempfile`

**Comment by henryiii** [comment]
Would it be okay to call git inside the unit test? Ideally, `git diff --exit-code` would actually be best here for the check.

**Comment by Czaki** [comment]
> Would it be okay to call git inside the unit test? Ideally, `git diff --exit-code` would actually be best here for the check.

Why? This will broke tests without creating commit. And I do not see option how it will help? Unit test does not call `build` function so will not create any files.

**Comment by henryiii** [comment]
@YannickJadoul or @joerick - I've generated a rather large number of runs, if you could cancel all bu the latest ones (at least in GHA), that would speed things up, I don't have permission to cancel a run (even if I started it, which I always thought would have been nice to be able to do)

@Czaki  The problem that I am trying to fix in this PR is that "build" *is* creating a file - causing a dirty git directory and breaking tools like `setuptools_scm`. It's not supposed to be creating any files *inside the git source directory*.

**Comment by henryiii** [comment]
@Czaki Ah, you are right, this is making a project that is not in git. Maybe we could compare the total number of files in the directory to some known total, then? Ignore my git check in the last commit, it will go away later ;)

**Comment by Czaki** [comment]
> @Czaki Ah, you are right, this is making a project that is not in git. Maybe we could compare the total number of files in the directory to some known total, then? Ignore my git check in the last commit, it will go away later ;)

class `TestProject` from `test/test_project/base.py` contains list of expected files in `files` variable. So it could be use to check list of files if all are listed in this var.

File: cibuildwheel/windows.py [hunk 1]
```diff
@@ -177,8 +177,14 @@ def pep_518_cp35_workaround(package_dir: Path, env: Dict[str, str]) -> None:
...
         )
+
         if requirements:
-            shell(['pip', 'install'] + requirements, env=env)
+            with tempfile.TemporaryDirectory() as d:
+                reqfile = Path(d) / "requirements.txt"
+                with reqfile.open("w") as f:
+                    for r in requirements:
+                        print(r, file=f)
+                call(['pip', 'install', '-r', str(reqfile)], env=env)
```
Comment (joerick) [inline cibuildwheel/windows.py:187]: `call()` should be able to take Path objects directly - this should work:

```suggestion
                call(['pip', 'install', '-r', reqfile], env=env)
```

File: test/test_pep518.py [hunk 2]
```diff
@@ -47,3 +49,8 @@ def test_pep518(tmp_path):
     # check that the expected wheels are produced
     expected_wheels = utils.expected_wheels("spam", "0.1.0")
     assert set(actual_wheels) == set(expected_wheels)
+
+    assert not (project_dir / "42").exists()
```
Comment (joerick) [inline test/test_pep518.py:53]: Might be worth adding a comment why these asserts are here. Feel free to link to #421 if it's easier

**Comment by YannickJadoul** [comment]
Also looks good to me, and approved by @joerick, so merging!

**Pull request merged**